Augmented Lagrangian method

aldma opened the pull request:
Dear JSOptimizers, this PR is to add an implementation of AL method for dealing with regularized problems with nonlinear (smooth) constraints.

aldma commented:
@dpo

aldma commented:
This implementation could take over the one in https://github.com/aldma/Bazinga.jl , which now has a slightly more flexible template but only a basic API.

dpo commented on src/AL_alg.jl:
```diff
@@ -0,0 +1,247 @@
+  nlp::AbstractNLPModel,
+  h::H,
+  options::ROSolverOptions;
+  subsolver = has_bounds(nlp) ? TRDH : PG,
```
I think better choices would be TR and R2 here. R2 is typically faster than PG. TRDH will only work if h is separable.

dpo commented on src/AL_model.jl:
```diff
@@ -0,0 +1,179 @@
+# based on Percival.jl
+# https://github.com/JuliaSmoothOptimizers/Percival.jl/blob/main/src/AugLagModel.jl
```
What are the differences with Percival's AugLagModel, if any?

dpo replied:
> we could include also the composition of the nonsmooth term with a smooth one.

What do you mean exactly? We would bundle the smooth term with the other smooth terms in order to keep the prox operators as simple as possible, wouldn't we?

aldma replied:
Say the cost is of the form ``f(x)+g(F(x))`` for some smooth `f`, `F` and nonsmooth `g`. An option is to reformulate the problem with some additional variables to avoid the composition of `g` with `F`. This approach works but there is also another option, which eliminates the auxiliary variables. Applying the AL machinery gives an AL function with a term being the Moreau envelope of `g` composed with a shifted `F`, and one always evaluates only the simple prox operator of `g`.
My point is, the AL function depends on the problem formulation and there are several approaches when some `g` appears in the problem.

dpo reviewed (commented):
Hi @aldma ! Many, many thanks for this! I just have a couple of very general comments for now, and a question: how does this solver compare to Bazinga?

aldma commented:
Bazinga can take problems with constraints in the form ``c(x) \in D``, where set ``D`` is easy to project onto. The ``AL`` solver here can handle these problems at the price of some additional auxiliary variables.
In terms of practical performance, I do not have any comparisons yet.

dpo commented:
@MohamedLaghdafHABIBOULLAH @MaxenceGollier Could you please make a pass on this code too? Thanks.

dpo commented on src/AL_alg.jl:
```diff
@@ -0,0 +1,250 @@
+
+At each iteration, a step s is computed as an approximate solution of
+
+    min  ½ ‖J(x) s + F(x)‖² + ½ σ ‖s‖² + ψ(s; x)
```
This is inaccurate, right?

aldma replied:
Right. Fixed with 829ceab

MaxenceGollier reviewed (commented):
Thank you @aldma for this PR !

Just a general comment, we are currently trying to revamp this repository to be more compliant with the JSO style of solvers (see #139). 
A few tips for your implementation
- Avoid using `options::ROSolversOptions` and use documented keyword arguments instead. this makes readibility easier for everyone as some solvers use the same variable names for different things.
- Add a solver structure (e.g `mutable struct nameSolver` where name is the name of your solver) that contains makes all the neccessary allocations for your solver when you construct it. The constructor should take an `AbstractNLPModel` as argument.
- Write the implementation of the solver in the function `SolverCore.solve!(nameSolver,...)` that allocates as less as possible (ideally it should not be allocating at all).
- Add a callback that is called at each iteration

If you need help with this, don't minding sending me messages. It would be very helpful for us. From my understanding, your implementation is similar to Percival.jl maybe you can look at how it is done there and try to work out something similar.

aldma commented on src/AL_model.jl:
```diff
@@ -0,0 +1,179 @@
+# based on Percival.jl
+# https://github.com/JuliaSmoothOptimizers/Percival.jl/blob/main/src/AugLagModel.jl
```
This file has been removed with 3d504cf. Now there is only a (default) method for the dual safeguard. The dependency on Percival could be weakened in the future.

aldma commented:
Thanks @MaxenceGollier for your feedback! I have revamped the AL solver following your instructions:
- removed `options::ROSolversOptions` and use now the unified kwargs.
- build `ALSolver` structure and implemented `solve!` method.
- added a callback.

Notice I added one method to `R2` too.

MaxenceGollier commented:
This looks great ! Thank you @aldma, 
Could you add unit tests to see whether `solve!` allocates ? 
Something similar to #147 perhaps ?

dpo commented:
@MaxenceGollier I’m assigning you to this PR. @aldma will not be able to take care of it in the next few months. Would you please check allocations and add unit tests? Thank you!

MaxenceGollier commented:
No problem, will look into it.

dpo commented:
Many thanks @aldma !
